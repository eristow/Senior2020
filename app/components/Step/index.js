/**
 *
 * Step
 *
 */

import styled from 'styled-components';
// import PropTypes from 'prop-types';

const Step = styled.button`
  margin-bottom: 1em;
  margin-left: 0.35em;
  margin-right: 0.35em;
  padding: 0em;
  width: 4em;
  height: 4em;
  font-size: 10px;
`;

Step.propTypes = {};

export default Step;
