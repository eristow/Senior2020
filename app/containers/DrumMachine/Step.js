import React from 'react';
import styled, { keyframes, css } from 'styled-components';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';
import PropTypes from 'prop-types';

import { makeSelectStepState } from './selectors';
import { changeSteps } from './actions';

const flash = keyframes`
  0% {
    opacity: 0.5;
  }

  50% {
    opacity: 1;
  }
  100% {
    opacity: 0.5
  }
`;

const flashMixin = css`
  animation: ${flash} 0.5s linear infinite;
`;

const StepButton = styled.button`
  flex: 1;
  background: ${props => (props.offsetColor ? '#25CCF7' : '#FD7272')};
  opacity: ${props => (props.on ? 1 : 0.35)};
  border-radius: 2;
  margin: 2px;
  ${props => props.doubled && flashMixin}
`;

const isOffsetColor = index =>
  (index > 3 && index < 8) || (index > 11 && index < 16);

export const Step = React.memo(
  ({ on, index, name, doubled, setSteps, stepState }) => {
    const toggleStep = e => {
      const shiftEnabled = e.shiftKey === true;
      const steps = [...stepState[name]];
      const val =
        steps[index] === 0
          ? shiftEnabled
            ? 2
            : 1
          : shiftEnabled && steps[index] === 1
          ? 2
          : 0;
      steps[index] = val;
      setSteps({
        ...stepState,
        [name]: steps,
      });
    };
    return (
      <StepButton
        on={on}
        offsetColor={isOffsetColor(index)}
        doubled={doubled}
        onClick={toggleStep}
      />
    );
  },
);

Step.propTypes = {
  on: PropTypes.bool,
  index: PropTypes.number,
  name: PropTypes.string,
  doubled: PropTypes.bool,
  setSteps: PropTypes.func,
  stepState: PropTypes.object,
};

const mapStateToProps = createStructuredSelector({
  stepState: makeSelectStepState(),
});

const mapDispatchToProps = dispatch => ({
  setSteps: value => {
    dispatch(changeSteps(value));
  },
});

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default compose(withConnect)(Step);
