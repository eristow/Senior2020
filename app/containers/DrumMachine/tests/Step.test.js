import React from 'react';
import { render } from 'react-testing-library';
import { Provider } from 'react-redux';
import configureStore from '../../../configureStore';

import { Step } from '../Step';

describe('<Step />', () => {
  let store;

  beforeAll(() => {
    store = configureStore();
  });

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    const dispatch = jest.fn();

    render(
      <Provider store={store}>
        <Step dispatch={dispatch} />
      </Provider>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <Step />
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });
});
