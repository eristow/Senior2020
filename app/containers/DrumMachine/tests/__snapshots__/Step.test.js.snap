// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Step /> Should render and match the snapshot 1`] = `
<button
  class="Step__StepButton-m11x8-0 epBtBi"
/>
`;
